src/hooks/useMedia.ts

```diff
@@ -0,0 +1,64 @@
+import { useEffect, useState } from "react";
+
+type UseMediaType = {
+    cameraList: InputDeviceInfo[];
+    micList: InputDeviceInfo[];
+    setCamera: React.Dispatch<React.SetStateAction<InputDeviceInfo | undefined>>;
+    setMic: React.Dispatch<React.SetStateAction<InputDeviceInfo | undefined>>;
+    selectedCamera: InputDeviceInfo | undefined;
+    selectedMic: InputDeviceInfo | undefined;
+    mediaStream: MediaStream | null;
+}
+
+const useMedia: () => UseMediaType = () => {
+    const [cameraList, setCameraList] = useState<InputDeviceInfo[]>([]);
+    const [micList, setMicList] = useState<InputDeviceInfo[]>([]);
+    
+    const [selectedCamera, setCamera] = useState<InputDeviceInfo>();
+    const [selectedMic, setMic] = useState<InputDeviceInfo>();
+
+    const [mediaStream, setMediaStream] = useState<MediaStream | null>(null);
+
+    useEffect(() => {
+        getMediaDevicesList();
+        
+        navigator.mediaDevices.getUserMedia({audio:true,video:true})
+        .then((stream) => {
+            stream.getTracks()
+            setMediaStream(stream)
+        })
+    }, [selectedCamera, selectedMic])
+    
+    const getMediaDevicesList = () => {
+        navigator.mediaDevices.enumerateDevices()
+        .then(devices => {
+            devices.forEach(device => {
+                device.kind === 'audioinput' && setMicList((currentList) => addUniqueDevice(currentList, device));
```

Comment:
should not call `setMicList` in each iteration, you can only add the device to the list in each iteration. Then call `setMicList` when finishing the loop.